processor/lsmintervalprocessor/internal/merger/model.go

```diff
@@ -406,3 +417,125 @@ func mergeDeltaHistogramDP(from, to pmetric.HistogramDataPoint) {
+
+	if to.Offset() > from.Offset() {
+		// Expand `to` to accomodate the merged buckets with offset as `from.Offset()`
+		newBucketsSize := max(fromSize, toSize) - int(from.Offset())
+		newBuckets := pcommon.NewUInt64Slice()
+		newBuckets.Append(make([]uint64, newBucketsSize)...)
+
+		// Copy `to` to the new bucket
+		for i := 0; i < toBuckets.Len(); i++ {
+			newBuckets.SetAt(int(to.Offset())+i, toBuckets.At(i))
+		}
+		newBuckets.MoveTo(to.BucketCounts())
+		to.SetOffset(from.Offset())
+
+		toBuckets = to.BucketCounts()
+	} else if fromSize > toSize {
+		toBuckets.Append(make([]uint64, fromSize-toSize)...)
+	}
+	for i := 0; i < fromBuckets.Len(); i++ {
+		toOffset := int(from.Offset()-to.Offset()) + i
+		toBuckets.SetAt(toOffset, toBuckets.At(toOffset)+fromBuckets.At(i))
+	}
+}
+
+// maybeScaleDownHistogram attempts to scale down histgram if the proper scale is passed.
+func maybeScaleDownHistogram(hist pmetric.ExponentialHistogramDataPoint, newScale int32) {
+	diff := hist.Scale() - newScale
+	switch {
+	case diff < 0:
+		panic("unexpected state, histogram cannot be upscaled")
+	case diff == 0:
+		return
+	}
+
+	scaleDownHistogramBuckets(hist.Positive(), diff)
+	scaleDownHistogramBuckets(hist.Negative(), diff)
+}
+
+// scaleDownHistogramBuckets scales down the histogram buckets by collapsing 2^(scaleDiff)
+// number of buckets into one bucket while handling the offset as required.
+// TODO (lahsivjar): Currently we don't trim the trailing zero entries from the buckets.
+// This is because OTel doesn't expose the data-model to efficiently slice the buckets,
+// instead, trimming trailing zero entries will require reallocation of the whole slice.
+// Is it worth it? (See test cases)
+func scaleDownHistogramBuckets(hist pmetric.ExponentialHistogramDataPointBuckets, scaleDiff int32) {
+	if scaleDiff == 0 {
+		return
+	}
+
+	// collapse is the number of buckets to collapse into lower scale bucket
+	// per higher scale bucket for downscaling
+	collapse := int64(1) << scaleDiff
+	buckets := hist.BucketCounts()
+
+	// higher represents the index position in the buckets count slice for higher
+	// scale histogram and lower represents the same for lower scale histogram
+	var higher, lower int
+	for higher < buckets.Len() {
+		// count represents number of higher buckets that will be collpased
+		// depending on the current index (may differ from `collapse` due to offset)
+		count := (int64(hist.Offset()) + int64(higher)) % collapse
+		for i := int64(count); i < collapse && higher < buckets.Len(); i++ {
+			if lower != higher {
+				buckets.SetAt(lower, buckets.At(higher)+buckets.At(lower))
+				buckets.SetAt(higher, 0)
+			}
+			higher++
+		}
+		lower++
+	}
+	hist.SetOffset(hist.Offset() >> scaleDiff)
+}
+
+func min[V constraints.Integer | constraints.Float](a, b V) V {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func max[V constraints.Integer | constraints.Float](a, b V) V {
+	if a > b {
+		return a
+	}
+	return b
+}
```

Comment:
Nit: use `cmp.Ordered` and `cmp.Less` instead of the experimental constraints package?

Reply:
Better yet use min and max builtin from the standard library https://pkg.go.dev/builtin#min, they are available since 1.21

Reply:
Yes! I thought there were functions, just couldn't remember where they were.

Reply:
> Better yet use min and max builtin from the standard library https://pkg.go.dev/builtin#min, they are available since 1.21

wth, I spent time trying to find these and failed 🤦

Reply:
Fixed, thanks for this!! I don't know why I remembered it being in some other package and not in standard library 🤦